embedded-graphics/src/primitives/ellipse.rs

```diff
@@ -0,0 +1,512 @@
+            threshold: 0,
+        }
+    }
+}
+
+impl Iterator for Points {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for point in &mut self.iter {
+            if is_point_inside_ellipse(self.size_sq, point * 2 - self.center_2x, self.threshold) {
+                return Some(point);
+            }
+        }
+
+        None
+    }
+}
+
+/// Pixel iterator for each pixel in the ellipse border
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    iter: Points,
+    outer_color: Option<C>,
+    inner_size_sq: Size,
+    inner_color: Option<C>,
+    center: Point,
+    threshold: u32,
+}
+
+impl<C> StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    fn new(styled: &Styled<Ellipse, PrimitiveStyle<C>>) -> Self {
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = styled.style.effective_stroke_width();
+
+        let outer_size = styled.primitive.size;
+        let inner_size = outer_size.saturating_sub(Size::new(2 * stroke_width, 2 * stroke_width));
+        let inner_size_sq = Size::new(inner_size.width.pow(2), inner_size.height.pow(2));
+        let iter = if !styled.style.is_transparent() {
+            Points::new(&styled.primitive)
+        } else {
+            Points::empty()
+        };
+
+        let threshold = compute_threshold(inner_size_sq, inner_size.width);
+
+        Self {
+            iter,
+            outer_color: styled.style.stroke_color,
+            inner_size_sq,
+            inner_color: styled.style.fill_color,
+            center: styled.primitive.center_2x(),
+            threshold,
+        }
+    }
+}
+
+impl<C> Iterator for StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for point in &mut self.iter {
+            let inside_border = is_point_inside_ellipse(
+                self.inner_size_sq,
+                point * 2 - self.center,
+                self.threshold,
+            );
+
+            let color = if inside_border {
+                self.inner_color
+            } else {
+                self.outer_color
+            };
+
+            if let Some(color) = color {
+                return Some(Pixel(point, color));
+            }
+        }
+
+        None
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_ellipse(self)
+    }
+}
+
+fn compute_threshold(size: Size, diameter: u32) -> u32 {
+    let Size {
+        width: a,
+        height: b,
+    } = size;
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        if diameter <= 4 {
+            diameter.pow(2) - diameter / 2
+        } else {
+            diameter.pow(2)
+        }
```

Comment:
We could share this code with `circle.rs` by making `diameter_to_threshold` in `circle.rs` `pub(crate)`. I'm not sure if that's an huge improvement, but it would make sure that they stay equal.

Reply:
Good idea. I went with `pub(in crate::primitives)` in `circle.rs` to keep it more tightly scoped.